Repository refactor

njmei opened the pull request:
This PR encompasses work to improve the usability and formatting of utilities and example notebooks for accessing AWS open datasets for the Allen Mouse Brain Atlas and Ivy Glioblastoma Atlas Project.

Improvements:
- The Ivy GAP example notebook should no longer cause crashes (please test this out if you have time @wbwakeman)
- The Allen Mouse Brain Atlas utils and example notebook no longer needs credentials to access data significantly simplifying access

TODO:
- [x] Need to fix existing tests for Allen Mouse Brain Atlas utils, they're currently all failing
- [x] (maybe ?) Need to get rid of all `os.path` stuff in favor of `pathlib.Path`
- [ ] ~(maybe ?) Need to add tests for new Ivy GAP utils~ (Not for this PR)
- [x] Need to update main repo readme
- [x] Change `https://ivygap.swedish.org/` url link in example notebook to  `https://ivygap.org/`

This PR is intended to address #1

njmei committed:
Rearrange repository structure

1) Move example notebooks to example_notebooks directory
   and standardize their naming

2) Move util.py files to package location

3) Remove /tmp directory and contents

njmei committed:
Make open_dataset_tools an installable package

njmei committed:
Refactor Ivy Glioblastoma Atlas Project example notebook and utils

njmei committed:
Refactor Allen Mouse Brain Atlas example notebook and utils

wbwakeman commented:
I think the URL of the "companion clinical database" still needs to be updated to: https://ivygap.org/
(At least that is the URL that seems to work now - https://ivygap.swedish.org/ does not - seems like that changed????)

wbwakeman commented:
GBM notebook downloads primary, annotation, boundary images but doesn't display them - is that expected?

![image](https://user-images.githubusercontent.com/1855315/121932710-07d19c80-ccfa-11eb-9d8f-a5e97fc8de6a.png)

njmei committed:
Improve _download_img() method for SectionDataSet

Previously this method was using mkstemp to create both
a file object handle as well as a filename then immediately
closing the file object. A more straightforward strategy is
to use a NamedTemporaryFile and 'with' context to
accomplish the same thing.

Also add more type annotations and regenerate the relevant notebook.

njmei committed:
Add matplotlib inline command

This may help resolve cases where images don't show up when
trying to run notebooks.

njmei committed:
Update website link for IVY GAP in notebook

njmei committed:
Fix existing tests for allen mouse brain atlas utils

njmei committed:
Update main repository README

wbwakeman commented:
Cool.  Yes, I see the images now. Thanks @njmei

danielsf commented on src/open_dataset_tools/ivy_gap_utils.py:
```diff
@@ -0,0 +1,309 @@
+    return load_s3_json_as_dataframe(client, bucket, key)
+
+
+class ImagePromise(object):
```
Do you mind adding a short docstring, in case someone else has to maintain this, so we know what this class is for?

wbwakeman replied:
Images look okay for me when I run that

danielsf commented on src/open_dataset_tools/ivy_gap_utils.py:
```diff
@@ -0,0 +1,309 @@
+
+            # If user hasn't provided a local_save_directory then assume
+            # we need to download from S3
+            if local_save_directory is None:
```
Do you need this if/else block? Doesn't `ImagePromise.__init__` default to `local_save_directory=None` (which, I think is the only difference between `if` and `else`?

danielsf reviewed (changes requested):
I think something got changed in the Allen Brain Atlas utils. If I run the notebook, the first image (around cell 19) looks like this
<img width="752" alt="first_coronal_image" src="https://user-images.githubusercontent.com/1682854/122869363-021f1c80-d2e1-11eb-913c-82abdde43625.png">
If I go back to the state of the repository before this branch, I do not see that large chunk of black at the bottom. Selecting other levels of downsampling for this sub image also results in black bars at different places in the image.

I'm not sure how that is even possible, given that the code claims to verify the MD5 checksum of the image on download.

This doesn't implicate the IVY blastoma code. If we want to merge this PR and take a new ticket to fix the ABA part of the repository (which, after all, is technically my problem), I am okay with that decision (@wbwakeman )

njmei commented:
@danielsf That is very bizarre... when I ran the notebook, the images looked normal:

https://github.com/AllenInstitute/open_dataset_tools/blob/repository-refactor/example_notebooks/Accessing_Allen_Mouse_Brain_Atlas_Data.ipynb

Though maybe there is a boto3 related download problem? I *did* end up needing to set: `ImageFile.LOAD_TRUNCATED_IMAGES = True` (See: https://pillow.readthedocs.io/en/stable/reference/ImageFile.html#PIL.ImageFile.LOAD_TRUNCATED_IMAGES)

danielsf commented:
@njmei I tried the notebook again this morning. The problem went away. Given the way we keep talking about this repository, I assume we don't want to invest the time in trying to figure out what happened last night. I'll approve the PR and we can wait to see if anyone else encounters this problem....?

danielsf reviewed (approved):
There were a few small in-line comments. Nothing major.

njmei committed:
Add documentation to ImagePromise class
